File: src/fracDimBatch.m
% fracDimBatch Calculates the fractal dimension of image (thresholding + edge detection)
% Copyright (c) 2016 Cai Jiashen

function [fd, thresI, bcFig] = fracDimBatch(cropI,binaryMask,sigma,epsilon,fdMode)

mag = 100;

%% Level Set Thresholding
% Sets level-set thresholding parameters if not set
if nargin <3
    sigma=4;
    epsilon=3;
end 

% Level set threshold
prethresI = regionfill(cropI,~binaryMask);
%prethresI = roifill(cropI,~binaryMask);

%figure();
%imshow (prethresI);

%prethresI = cropI;

thresI = ~levelSetBatch(prethresI,sigma,epsilon);

thresI=thresI.*binaryMask;

%% Check FD Mode
if strcmp(fdMode,'LV')
    largestAreaMode = 'JS'; % or 'PFT'
elseif strcmp(fdMode,'RV')
    largestAreaMode = 'RV';
end

%% Ellipse around largest area of thresholded image

if strcmp(largestAreaMode,'JS')
    % JS Original ConvexHull/Ellipse
    largestAreaI=bwconvhull(bwpropfilt(imbinarize(thresI,0.5),'Area',1,'largest'));
    statsLA = regionprops(largestAreaI,'Centroid', 'MajorAxisLength','MinorAxisLength','Orientation');
    
    hortLA = statsLA.MajorAxisLength/2;
    vertLA = statsLA.MinorAxisLength/2;
    
    xLA = statsLA.Centroid(1);
    yLA = statsLA.Centroid(2);
    
    [ySize, xSize] = size(largestAreaI);
    
    [xMask, yMask] = meshgrid(1:xSize,1:ySize);
    lAMask = ((xMask-xLA)/hortLA).^2 + ((yMask-yLA)/hortLA).^2 <= 1;
    
    compositeLAI=imadd(largestAreaI,lAMask);
    compositeLAI(compositeLAI>1)=1;

elseif strcmp(largestAreaMode,'PFT')
    % PFT Adaptation of ConvexJull/Ellipse + JS's edits/convention
    largestAreaI = bwconvhull(bwpropfilt(imbinarize(thresI, 0.5), 'Area', 1, 'largest'));
    
    statsLA = regionprops(largestAreaI, 'Centroid', 'MajorAxisLength', 'MinorAxisLength', 'Orientation');
    
    xLA = statsLA.Centroid(1);
    yLA = statsLA.Centroid(2);
    hortLA = 1.05*statsLA.MajorAxisLength/2.0;    % Expand the axis by 5 P.C.
    vertLA = 1.05*statsLA.MinorAxisLength/2.0;    % Expand the axis by 5 P.C.
    Theta = (pi/180.0)*statsLA.Orientation;
    cc = cos(Theta);
    ss = sin(Theta);
    
    [ySize, xSize] = size(largestAreaI);
    
    [ xx, yy ] = meshgrid(1:xSize, 1:ySize);
    
    XX = cc*(xx - xLA) - ss*(yy - yLA);
    YY = ss*(xx - xLA) + cc*(yy - yLA);
    
    equivEllipse = (XX.^2/hortLA^2 + YY.^2/vertLA^2 <= 1.0);
    
    delta = round(0.025*(hortLA + vertLA));
    largestAreaI = imdilate(largestAreaI, strel('disk', delta, 4));   % Expand the largest area by roughly the same amount as the ellipse
    
    compositeLAI = largestAreaI | equivEllipse;
elseif strcmp(largestAreaMode,'RV')
    
    % Boundary check for RV segmentation
    countPixels = nnz(thresI);
    invertThresI = (~thresI).*binaryMask;
    countInvertPixels = nnz(invertThresI);

    intensityPixels = sum(sum(cropI.*thresI))/countPixels;
    intensityInvertPixels = sum(sum(cropI.*invertThresI))/countInvertPixels;

    if (intensityInvertPixels>intensityPixels)
        thresI = invertThresI;
        disp('Flip Pixels');
    end
    
    % Largest area
    largestAreaI=bwconvhull(bwpropfilt(imbinarize(thresI,0.5),'Area',1,'largest'));
    delta = 2;
    largestAreaI = imdilate(largestAreaI, strel('disk', delta, 4));
    compositeLAI=largestAreaI;
end

thresI=thresI.*compositeLAI;

%% Edge detection
[~, threshold] = edge(thresI,'sobel');
edgeI = edge(thresI,'sobel',threshold*0.5);


%% Box-counting with FD computation
[fd, bcFig] = bxct(edgeI);

% Overview plots
% figure('Name','Overview','NumberTitle','off');
% subplot(2,2,1), imshow(cropI), title('Gray');
% subplot(2,2,2), imshow(cropI), title('Crop');%, hold on, contour (thresI,'w');
% subplot(2,2,3), imshow(cropI), title(sprintf('Threshold')), hold on, contour (thresI,'w');
% subplot(2,2,4), imshow(edgeI), title(sprintf('Edges, FD = %0.5f',fd));

File: python/src/fracDimBatch.py
import numpy as np
import cv2
from skimage.segmentation import flood_fill
from skimage.filters import sobel
from skimage.measure import label, regionprops
from skimage.morphology import convex_hull_image, dilation, disk
from skimage.feature import canny
import matplotlib.pyplot as plt

from levelSetBatch import level_set_batch

def level_set_thresholding(cropI, binaryMask, sigma=4, epsilon=3):
    print("Level set thresholding")
    if prethresI.shape != binaryMask.shape:
        raise ValueError(f"Shape mismatch: prethresI has shape {prethresI.shape}, but binaryMask has shape {binaryMask.shape}")
    prethresI = flood_fill(cropI, ~binaryMask.astype(bool), 0)
    thresI = level_set_batch(prethresI, sigma, epsilon)
    thresI = np.logical_not(thresI) * binaryMask
    return thresI

def compute_largest_area(thresI, fdMode):
    print("Computing largest area")
    if fdMode == 'LV':
        largestAreaMode = 'JS'
    elif fdMode == 'RV':
        largestAreaMode = 'RV'
    
    if largestAreaMode == 'JS':
        largestAreaI = convex_hull_image(thresI)
        stats = regionprops(largestAreaI.astype(int))[0]
        hortLA = stats.major_axis_length / 2
        vertLA = stats.minor_axis_length / 2
        xLA, yLA = stats.centroid
        ySize, xSize = largestAreaI.shape
        xMask, yMask = np.meshgrid(np.arange(xSize), np.arange(ySize))
        lAMask = ((xMask - xLA) / hortLA) ** 2 + ((yMask - yLA) / hortLA) ** 2 <= 1
        compositeLAI = np.logical_or(largestAreaI, lAMask)
        
    elif largestAreaMode == 'PFT':
        largestAreaI = convex_hull_image(thresI)
        stats = regionprops(largestAreaI.astype(int))[0]
        xLA, yLA = stats.centroid
        hortLA = 1.05 * stats.major_axis_length / 2
        vertLA = 1.05 * stats.minor_axis_length / 2
        theta = np.deg2rad(stats.orientation)
        cc, ss = np.cos(theta), np.sin(theta)
        ySize, xSize = largestAreaI.shape
        xx, yy = np.meshgrid(np.arange(xSize), np.arange(ySize))
        XX = cc * (xx - xLA) - ss * (yy - yLA)
        YY = ss * (xx - xLA) + cc * (yy - yLA)
        equivEllipse = (XX ** 2 / hortLA ** 2 + YY ** 2 / vertLA ** 2 <= 1.0)
        delta = round(0.025 * (hortLA + vertLA))
        largestAreaI = dilation(largestAreaI, disk(delta))
        compositeLAI = np.logical_or(largestAreaI, equivEllipse)
        
    elif largestAreaMode == 'RV':
        countPixels = np.count_nonzero(thresI)
        invertThresI = np.logical_not(thresI) * binaryMask
        countInvertPixels = np.count_nonzero(invertThresI)
        intensityPixels = np.sum(cropI * thresI) / countPixels
        intensityInvertPixels = np.sum(cropI * invertThresI) / countInvertPixels
        
        if intensityInvertPixels > intensityPixels:
            thresI = invertThresI
        largestAreaI = convex_hull_image(thresI)
        delta = 2
        largestAreaI = dilation(largestAreaI, disk(delta))
        compositeLAI = largestAreaI
        
    thresI = thresI * compositeLAI
    return thresI

def edge_detection(thresI):
    print("Edge detection")
    edgeI = canny(thresI, sigma=1)
    return edgeI

def box_counting(edgeI):
    print("Box counting")
    # Placeholder function for box-counting algorithm
    # Compute the fractal dimension here
    fd = 1.23  # Dummy value
    bcFig = None  # Placeholder for the box-counting figure
    return fd, bcFig

def fracDimBatch(cropI, binaryMask, sigma=4, epsilon=3, fdMode='LV'):
    print("Fractal dimension batch")
    thresI = level_set_thresholding(cropI, binaryMask, sigma, epsilon)
    thresI = compute_largest_area(thresI, fdMode)
    edgeI = edge_detection(thresI)
    fd, bcFig = box_counting(edgeI)
    return fd, thresI, bcFig
